### PR TITLE
Provide ArmiPlugin.defineSystemGrid builders

### PR DESCRIPTION
**Pull request by drewj-usnctech**
## Description

---
Provide ArmiPlugin.defineSystemGrid builders

Goal is to let a plugin define additional sections that
can reside in the ``systems`` section, e.g.,
```yaml
systems:
    core:
        grid name: core
        origin:
            x: 0
            y: 0
            z: 0
    sfp:
        grid name: spent fuel pool
        type: sfp
        ...
```

The end goal is to have a new system for control drums, pointing to a designated
grid, and allowing assemblies to be added to that grid rather than the core grid.

The plugins basically define a map from the system grid type (defaults to "core"
in the blueprint) and the builder function. The builder function will most likely
be the class of the thing to be built (e.g., ``"core"`` returns a ``Core`` **type**
and the blueprint construct method creates the instance given the name of the grid.
In the core example it's not super helpful. But for the spent fuel pool, the
``type`` is ``sfp`` but the name is defined by the key one level up and does not
have to match any specific criteria.

The `ReactorPlugin` defines the existing ``"core"`` and ``"sfp"`` builders, but is
tagged to go last (or near last) in the YAML construction sequence. This allows other
plugins to define their own core builders and let those have preference over the
default.

Signed-off-by: Andrew Johnson <a.johnson@usnc-tech.com>

## Checklist

- [ ] The code is understandable and maintainable to people beyond the author.
- [ ] Tests have been added/updated to verify that the new or changed code works.
- [ ] There is no commented out code in this PR.
- [ ] The commit message follows [good practices](https://terrapower.github.io/armi/developer/tooling.html).
- [ ] All docstrings are still up-to-date with these changes.

If user exposed functionality was added/changed:

- [ ] Documentation added/updated in the `doc` folder.
- [ ] New or updated dependencies have been added to `setup.py`.

### CODE REVIEW

File: armi/reactor/blocks.py [hunk 1]
```diff
@@ -188,7 +188,7 @@ def r(self):
 
         core = self.core
         if core is None:
-            return None
+            return self.getAncestor(lambda o: isinstance(o, Reactor))
```
Comment (drewj-usnctech) [inline armi/reactor/blocks.py:191]: This change only appeared for `Composites` that were not directly attached to the core, but
more acted as siblings to the `Core`. For example, if the spent fuel pool composite was
a child of `Reactor`, and the spent fuel pool contained assemblies which in turn contained blocks,
various HDF serialization methods would error out because Block.core -> None.

One example:
_addHomogenizedNumberDensityParams -> collectBlockNumberDensities ->  getSymmetryFactor
calls self.r.core.spatialGrid.overlapsWhichSymmetryLine

The docstring for Block.r indicates that most calls to self.r are to obtain the core
and may be better off just calling Block.core. But, here exists a case where a Block
may not have a Core as the parent (e.g., ex-core structures that physically reside
in the reactor) but still needs to access, for some reason, the spatial grid.

A deeper issue is that this call stack is trying to find the spatial grid for a block
assuming the block is on the Core.spatialGrid. Not sure how to square that. Maybe
the call for symmetry line tries to find a parent with a spatialGrid?

File: armi/plugins.py [hunk 2]
```diff
@@ -120,13 +120,16 @@
...
+from typing import Dict, Union, Callable, Optional, TYPE_CHECKING
 
 import pluggy
 
 from armi import pluginManager
 from armi.utils import flags
 
+if TYPE_CHECKING:
+    from armi.reactor.composites import Composite
+
```
Comment (drewj-usnctech) [inline armi/plugins.py:132]: Not sure the stance on type hinting but seems positive (#531  / #493). This block is only to help with an IDE showing a type that exists in the function signature, e.g.,
```python
def func(a: "Composite"):
```
does not error if `Composite` is imported in this way. But does if you remove the quotes

**Comment by drewj-usnctech** [comment]
@john-science @jakehader @ntouran any interest in this feature?

File: armi/plugins.py (hunk 2 above)
Comment (john-science) [reply armi/plugins.py:132]: We are definitely pro-type hinting. For sure.

I'm not thrilled about extra import and `TYPE_CHECKING`.  But perhaps this is the way of the future.  I haven't decided about these details yet.

Still, yes, generally, we like type hinting.

**Comment by john-science** [comment]
Sorry for not looking at this a long time ago.

I've been busy, and just ignore PRs with the red "X" next to them, but I see that's a minor code coverage thing here.

**Comment by john-science** [comment]
@ntouran Do you have an opinion?

File: armi/reactor/blocks.py (hunk 1 above)
Comment (john-science) [reply armi/reactor/blocks.py:191]: Perhaps this could use a one-line comment explaining why?

Comment (john-science) [reply armi/reactor/blocks.py:191]: @ntouran  I can't tell, will this be confusing to users?  They could ask for `b.r` and get something that is not the `Reactor`.

File: armi/reactor/blueprints/reactorBlueprint.py [hunk 3]
```diff
@@ -74,25 +74,29 @@ def __init__(self, name=None, gridName=None, origin=None):
...
+                # define a new core builder before finding those defined by the
+                # ReactorPlugin
+                if key == typ:
+                    return builder
+                seen.add(key)
+
+        raise ValueError(
+            "Could not determine an appropriate class for handling a "
+            "system of type `{}`. Supported types are {}.".format(typ, sorted(seen))
+        )
```
Comment (john-science) [inline armi/reactor/blueprints/reactorBlueprint.py:99]: @drewj-usnctech This is a non-trivial change.  And while it _looks_ fine to me, it is a bit hard for me to tell if there are edge cases where this would fail. What would help me understand/grok the changes if if we had a unit test to cover this new/re-written method. Does that make sense?

The obvious place would seem to be `armi/reactor/blueprints/tests/test_reactorBlueprints.py`.

Side Note: Again, I'm sorry it took so long for me to get to this PR.  I wasn't ignoring you, I've just been very busy and when I saw the little red "X" next to your PR I thought it wasn't ready yet.  So my triage procedure was faulty. My  bad.

Comment (drewj-usnctech) [reply armi/reactor/blueprints/reactorBlueprint.py:99]: @john-science you're totally fine on the delay. and yes, this is a non-trivial change where a unit test would help. I'm not sure how to handle a unit test where we would add something to the plugin manager returned by `armi.getPluginMnagerOrFail`. The testable way might be to make this rely on another method that, given any plugin manager, performs this lookup. Then we don't have to touch `armi.getPluginManagerOrFail`

Possible path: refactor the call signature to accept a string `typ` and the plugin manager, then test that "private" method? How do you feel about that?

File: armi/plugins.py (hunk 2 above)
Comment (drewj-usnctech) [reply armi/plugins.py:132]: Excellent. And by

> I'm not thrilled about extra import

Do you mean the import here of `from armi.reactor.composites import Composite`? If it's a performance issue you're considering, that import will not trigger in a production run (maybe the cause for a coverage drop) because of how python and type checkers like IDEs sets the `TYPE_CHECKING` variable

> A special constant that is assumed to be True by 3rd party static type checkers. It is False at runtime.

So then this import isn't tripped in e.g., `armi run`

https://docs.python.org/3.7/library/typing.html#typing.TYPE_CHECKING

File: armi/reactor/blocks.py (hunk 1 above)
Comment (drewj-usnctech) [reply armi/reactor/blocks.py:191]: > They could ask for b.r and get something that is not the Reactor.

That actually would have been the previous case if you had spun up a Reactor that contained a `Core` and some sibling `Composite` that also contained `Block` s and `Assembly` s. Because this sibling isn't the `Core`, blocks assigned have a non-None `self.core` attribute from
https://github.com/terrapower/armi/blob/a3f23ed01538aaaeb1dc58b5bf158def918901ef/armi/reactor/blocks.py#L165-L170

So here, for the blocks in the `Core`-adjacent `Composite`, their parent tree looks like `Block` -> `Assembly` -> `Sibling` -> `Reactor`. Then, if we did `b.core` we get `None,` which makes `b.r` return `None` without this change.

File: armi/plugins.py (hunk 2 above)
Comment (john-science) [reply armi/plugins.py:132]: Yeah, I was just thinking about performance. 

To be clear, your typing code here is fine. I am just tenuously unhappy with the general idea of having more imports in Python modules, just for the purposes of typing. I think the Python team's design decision seems likely to generally cause more imports and performance hits, to provide typing info. Which... is just insanely backwards.

Again, your code here is fine. I'm just not sure the languages typing solution is a good one.

**Comment by drewj-usnctech** [comment]
@john-science @ntouran thank you for the feedback. I added some docs in a place that I think makes sense, but let me know if there's a better way to present this information.

In adding the docs, I found some possible shortcomings related to the `sfp` system documentation. The [Systems blueprint documentation](https://github.com/terrapower/armi/blob/master/doc/user/inputs/blueprints.rst#systems) has content like
```yaml
systems:
    core:
        grid name: core
        origin:
            x: 0.0
            y: 10.1
            z: 1.1
    sfp:
        grid name: sfp
        origin:
            x: 1000.0
            y: 12.1
            z: 1.1
```
which should communicate to the user that a `SpentFuelPool` object should be created and attached to the `sfp` grid. But, because the `type` field is missing, another `Core` is built instead.

Demoing this with
```diff
diff --git a/armi/reactor/blueprints/tests/test_reactorBlueprints.py b/armi/reactor/blueprints/tests/test_reactorBlueprints.py
index 6d47f5a..39f002f 100644
--- a/armi/reactor/blueprints/tests/test_reactorBlueprints.py
+++ b/armi/reactor/blueprints/tests/test_reactorBlueprints.py
@@ -19,6 +19,7 @@ import os
 from armi.reactor import blueprints
 from armi import settings
 from armi.reactor import reactors
+from armi.reactor.assemblyLists import SpentFuelPool
 from armi.reactor.blueprints import reactorBlueprint
 from armi.reactor.blueprints import gridBlueprint
 from armi.reactor.blueprints.tests import test_customIsotopics
@@ -104,6 +105,7 @@ class TestReactorBlueprints(unittest.TestCase):
         core, sfp = self._setupReactor()
         self.assertEqual(len(core), 2)
         self.assertEqual(len(sfp), 4)
+        self.assertIsInstance(sfp, SpentFuelPool)
 
     def test_materialDataSummary(self):
         """Test that the material data summary for the core is valid as a printout to the stdout."""
```
yields an assertion error
```
Traceback (most recent call last):
  File "/Users/ajohnson/codes/armi/armi/reactor/blueprints/tests/test_reactorBlueprints.py", line 108, in test_construct
    self.assertIsInstance(sfp, SpentFuelPool)
  File "/opt/anaconda3/envs/raptor/lib/python3.7/unittest/case.py", line 1276, in assertIsInstance
    self.fail(self._formatMessage(msg, standardMsg))
  File "/opt/anaconda3/envs/raptor/lib/python3.7/unittest/case.py", line 693, in fail
    raise self.failureException(msg)
AssertionError: <Core: sfp id:140658844990032> is not an instance of <class 'armi.reactor.assemblyLists.SpentFuelPool'>
```

**Comment by drewj-usnctech** [comment]
@john-science ah that's on me. I thought, because the upstream main branch had been updated with the click changes, I wouldn't need anything on my end. But I've merged in main and it _should_ work now 🤞

**Comment by drewj-usnctech** [comment]
@john-science potentially very minor coverage drop. How do you want to proceed?

**Comment by john-science** [comment]
> @john-science potentially very minor coverage drop. How do you want to proceed?

No worries. It's a 4th decimal-place drop.  I'm not bothered.

**Comment by john-science** [comment]
> @john-science @ntouran thank you for the feedback. I added some docs in a place that I think makes sense, but let me know if there's a better way to present this information.
> 
> In adding the docs, I found some possible shortcomings related to the `sfp` system documentation. The [Systems blueprint documentation](https://github.com/terrapower/armi/blob/master/doc/user/inputs/blueprints.rst#systems) has content like
> 
> ```yaml
> systems:
>     core:
>         grid name: core
>         origin:
>             x: 0.0
>             y: 10.1
>             z: 1.1
>     sfp:
>         grid name: sfp
>         origin:
>             x: 1000.0
>             y: 12.1
>             z: 1.1
> ```
> 
> which should communicate to the user that a `SpentFuelPool` object should be created and attached to the `sfp` grid. But, because the `type` field is missing, another `Core` is built instead.
> 
> Demoing this with
> 
> ```diff
> diff --git a/armi/reactor/blueprints/tests/test_reactorBlueprints.py b/armi/reactor/blueprints/tests/test_reactorBlueprints.py
> index 6d47f5a..39f002f 100644
> +++ b/armi/reactor/blueprints/tests/test_reactorBlueprints.py
> @@ -104,6 +105,7 @@ class TestReactorBlueprints(unittest.TestCase):
>          core, sfp = self._setupReactor()
>          self.assertEqual(len(core), 2)
>          self.assertEqual(len(sfp), 4)
> +        self.assertIsInstance(sfp, SpentFuelPool)
>  
> ```
> 
> yields an assertion error
> 
> ```
> Traceback (most recent call last):
>   File "/Users/ajohnson/codes/armi/armi/reactor/blueprints/tests/test_reactorBlueprints.py", line 108, in test_construct
>     self.assertIsInstance(sfp, SpentFuelPool)
>   File "/opt/anaconda3/envs/raptor/lib/python3.7/unittest/case.py", line 1276, in assertIsInstance
>     self.fail(self._formatMessage(msg, standardMsg))
>   File "/opt/anaconda3/envs/raptor/lib/python3.7/unittest/case.py", line 693, in fail
>     raise self.failureException(msg)
> AssertionError: <Core: sfp id:140658844990032> is not an instance of <class 'armi.reactor.assemblyLists.SpentFuelPool'>
> ```

So, this is the last question in this PR?  Thoughts? @ntouran ?
